### DISCUSSION TITLE
[JUnit Test] Test failed 5 March 2023

### DISCUSSION THREAD

**Discussion by wolfgang-ch** [mytourbook/mytourbook#1006] [General]
This is the build error report

[Tests-Failures.zip](https://github.com/mytourbook/mytourbook/files/10891485/Tests-Failures.zip)

![image](https://user-images.githubusercontent.com/1283445/222952642-a94d8bdb-68d5-4e33-9327-375d977a33b0.png)

I'm struggling with these many errors when I run the test on my PC ...

![image](https://user-images.githubusercontent.com/1283445/222952419-143d8d2f-06a4-4449-9478-bc855494b2fd.png)

... and when I run the test on my 2nd PC with I use to develop the 2.5D animation, I get these errors which are also in the above `Tests-Failures.zip`

[junit-test-error.txt](https://github.com/mytourbook/mytourbook/files/10891477/junit-test-error.txt)

I change the mnemonic for `Duplicate Tour...` which may cause a unit test error

![image](https://user-images.githubusercontent.com/1283445/222953039-a6dd57db-d53f-4cb8-bf03-bfa515982f4b.png)

**Comment by FJBDev** [comment]
> I'm struggling with these many errors when I run the test on my PC ...

It can be hard to find the actual test that fails because for some tests, the SWTBot framework will fail all the tests that run after the failed tests.
But I added at the time a feature that makes it easier:
For each failed build, you will see a zip archive that contains information why the unit tests failed. For the UI unit tests, it contains a screenshot where the tests failed and stopped.
![image](https://user-images.githubusercontent.com/25235665/223007009-92e1a63d-064e-48bf-aea8-8ba4dafee3ab.png)


In this case, it's because this message was added and the Unit tests now need to click on this dialog
![org junit jupiter engine descriptor MethodExtensionContext testDuplicateAndDeleteTour](https://user-images.githubusercontent.com/25235665/223006888-299dba28-3e4b-4228-90a3-483a55ec1d34.png)


> ... and when I run the test on my 2nd PC with I use to develop the 2.5D animation, I get these errors which are also in the above `Tests-Failures.zip`

Did you make sure to run the tests as [JUnit Plugin Test](https://github.com/mytourbook/mytourbook/pull/252)

> I change the mnemonic for `Duplicate Tour...` which may cause a unit test error
> 

This is fixed in [this PR](https://github.com/mytourbook/mytourbook/pull/1007)

**Reply by wolfgang-ch** [reply]
>But I added at the time a feature that makes it easier: For each failed build, you will see a zip archive that contains information why the unit tests failed. For the UI unit tests, it contains a screenshot where the tests failed and stopped.

I opened this zip archive, looked into the .log file but didn't opened the image file because I didn't knew that this is the important one 😄

**Reply by wolfgang-ch** [reply]
> > ... and when I run the test on my 2nd PC with I use to develop the 2.5D animation, I get these errors which are also in the above `Tests-Failures.zip`
> 
> Did you make sure to run the tests as [JUnit Plugin Test](https://github.com/mytourbook/mytourbook/pull/252)

~~When I start the JUnit Plugin Test from the context menu then it works but issues occure when starting with a toolbar action which runs the previous test even when I cleaned the workspace, strange~~

Yesterday I was may be a bit confused but below I found the reason why it failed

**Reply by wolfgang-ch** [reply]
After testing again, I found this issue:

This code fails

![image](https://user-images.githubusercontent.com/1283445/223040425-c24ea740-130f-4516-9d98-e38162501424.png)

because the node text is "2", when setting "2" into the code then the test of this part is OK

![image](https://user-images.githubusercontent.com/1283445/223040508-c10cd07d-2743-48eb-a376-6687a5e90941.png)

```
org.eclipse.swtbot.swt.finder.exceptions.WidgetNotFoundException: Could not find node with text: 1
	at org.eclipse.swtbot.swt.finder.widgets.SWTBotTreeItem.getNodes(SWTBotTreeItem.java:337)
	at org.eclipse.swtbot.swt.finder.widgets.SWTBotTreeItem.getNode(SWTBotTreeItem.java:310)
	at org.eclipse.swtbot.swt.finder.widgets.SWTBotTreeItem.getNode(SWTBotTreeItem.java:349)
	at dialogs.DialogExtractToursTests.extractTour(DialogExtractToursTests.java:63)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
```

**Reply by FJBDev** [reply]
> I opened this zip archive, looked into the .log file but didn't opened the image file because I didn't knew that this is the important one 😄

I am glad I added the image because "a picture is worth 100 words" as I find it helps more than looking at a long log file

**Reply by FJBDev** [reply]
> This code fails
> because the node text is "2", when setting "2" into the code then the test of this part is OK
> 

Is it failing on your machine ? If it is, then it might be a timezone issue. I believe that setting [this date](https://github.com/mytourbook/mytourbook/blob/6a90cc854bc09805d0e68daf414f6d1f862575e7/bundles/net.tourbook.ui.tests/src/test/dialogs/DialogExtractToursTests.java#L56) will result in a different day for the extracted tour depending on your machine's timezone.
At this moment, some tests will fail if your PC is on a different timezone than UTC. It's something I meant to fix but haven't finished yet (probably one of my next PR), especially since I didn't know other people than me and GitHub were running tests 😄

**Reply by FJBDev** [reply]
> I opened this zip archive, looked into the .log file but didn't opened the image file because I didn't knew that this is the important one 😄

The image is very useful and even the name is very helpful because it contains the name of the test for which it failed

![image](https://user-images.githubusercontent.com/25235665/223281025-7ad1eafb-899c-4ef0-9336-c73cfdd8a8ed.png)

**Comment by wolfgang-ch** [comment]
When running the unit test, it is always asking to update the database when the git repo is reset

![image](https://user-images.githubusercontent.com/1283445/223139622-3016e65f-dcf0-46d4-8aef-01b3dc74c15a.png)

**Reply by FJBDev** [reply]
Yes, and that is why I added this argument to not display it and instead "update silently". It is used by default when ran in the Github build [here](https://github.com/mytourbook/mytourbook/blob/6a90cc854bc09805d0e68daf414f6d1f862575e7/bundles/net.tourbook.ui.tests/pom.xml#LL46C98-L46C120) and you can use it to not have this dialog displayed

**Reply by wolfgang-ch** [reply]
This parameter worked 😄 

![image](https://user-images.githubusercontent.com/1283445/223338901-31dc9d37-58bf-484c-a833-7f1e866fbc0a.png)

**Comment by wolfgang-ch** [comment]
Another issue: 

Sometimes, not always, the WMS dialog is not filled out with the provider url

As I remember, the error message was, that it cannot find the widget

![image](https://user-images.githubusercontent.com/1283445/223340578-9f7a475c-0022-4d1c-8b30-f10c13e7f9a0.png)

![image](https://user-images.githubusercontent.com/1283445/223340498-43cd6948-c85d-4f77-9d83-a86a47688a67.png)

**Reply by FJBDev** [reply]
Hmm, weird. 
Did you run it in Debug mode to see why it is not filled ? And which error stack trace did you get ?

**Reply by wolfgang-ch** [reply]
The next time when it occurs I attach the stack trace here, since then I couldn't reproduce it

**Comment by wolfgang-ch** [comment]
Is there an automatic cleanup procedure for the test workspace, currently I have to manually ...

* Rest the git repo
* Delete new files

... before I start a new test

**Reply by FJBDev** [reply]
> Is there an automatic cleanup procedure for the test workspace, currently I have to manually ...
> 
>     * Rest the git repo
> 
>     * Delete new files
> 
> 
> ... before I start a new test

There is no automatic procedure as of now but below is what I do:
- I delete the `workspace` folder and replace it with the [original one](https://github.com/mytourbook/mytourbook/tree/main/bundles/net.tourbook.ui.tests/workspace)

**Reply by wolfgang-ch** [reply]
> There is no automatic procedure as of now but below is what I do:
> 
>     * I delete the `workspace` folder and replace it with the [original one](https://github.com/mytourbook/mytourbook/tree/main/bundles/net.tourbook.ui.tests/workspace)

I hate to do repeated tasks, so I've written an ant script to reset the workspace https://github.com/mytourbook/mytourbook/commit/ca007069e85048b0ef61564143aa6b71acc9f136

![image](https://user-images.githubusercontent.com/1283445/223653033-e749ecd2-a9c8-4864-9964-23676fc6d405.png)

You have to define these arguments once

![image](https://user-images.githubusercontent.com/1283445/223653097-f0ddeda9-3675-4235-876f-19df1da4d541.png)

**Reply by FJBDev** [reply]
> I hate to do repeated tasks, so I've written an ant script to reset the workspace [ca00706](https://github.com/mytourbook/mytourbook/commit/ca007069e85048b0ef61564143aa6b71acc9f136)
> 
> ![image](https://user-images.githubusercontent.com/1283445/223653033-e749ecd2-a9c8-4864-9964-23676fc6d405.png)
> 
> You have to define these arguments once
> 
> ![image](https://user-images.githubusercontent.com/1283445/223653097-f0ddeda9-3675-4235-876f-19df1da4d541.png)

Nice, I am glad my unit tests are getting contribution.

I am not familiar with ant tasks, how do you specify it to get run after all the UI tests are done ?

**Reply by wolfgang-ch** [reply]
The ant script has still to be run manually from the external tools menu. The second time you have to press only the action button not the menu action until you run another external tool.

![image](https://user-images.githubusercontent.com/1283445/223934198-634d8800-aa78-42b9-9d54-606c79639690.png)

**Reply by wolfgang-ch** [reply]
I just discovered this: You have to ensure that the MT git repo is selected, otherwise the ant is running in the wrong repo with this error

![image](https://user-images.githubusercontent.com/1283445/223935325-5f506c17-7e22-494b-9ba9-448f9be9e0f5.png)

**Reply by FJBDev** [reply]
Thank you for this ant script that is very valuable in saving time and not do manual actions. I had never done or used Ant scripts before
